controller/tenant.go

```diff
@@ -337,6 +298,74 @@ func listNames(namespaces []*tenant.Namespace) []string {
 	return names
 }
 
+func (c *TenantController) getOrInitTenant(ctx context.Context, user *auth.User) (*tenant.Tenant, error) {
+	var dbTenant *tenant.Tenant
+	var err error
+	tenantRepository := c.tenantService.NewTenantRepository(user.ID)
+	// check if tenant already exists
+	if tenantRepository.Exists() {
+		dbTenant, err = c.getExistingTenant(ctx, user.ID, user.OpenShiftUsername)
+		if err != nil {
+			log.Error(ctx, map[string]interface{}{
+				"err":      err,
+				"tenantID": user.ID,
+			}, "retrieval of tenant entity from DB failed")
+			return nil, errors.NewNotFoundError("tenant", user.ID.String())
+		}
+	} else {
+		nsBaseName, err := tenant.ConstructNsBaseName(c.tenantService, environment.RetrieveUserName(user.OpenShiftUsername))
+		if err != nil {
+			log.Error(ctx, map[string]interface{}{
+				"err":         err,
+				"os_username": user.OpenShiftUsername,
+			}, "unable to construct namespace base name")
+			return nil, errors.NewInternalError(ctx, err)
+		}
+		// if does not exist then create a new tenant
+		dbTenant = &tenant.Tenant{
+			ID:         user.ID,
+			Email:      *user.UserData.Email,
+			OSUsername: user.OpenShiftUsername,
+			NsBaseName: nsBaseName,
+		}
+		err = tenantRepository.CreateTenant(dbTenant)
+		if err != nil {
+			log.Error(ctx, map[string]interface{}{
+				"err": err,
+			}, "unable to store tenant configuration")
+			return nil, err
+		}
+	}
+	return dbTenant, nil
+}
+
+func (c *TenantController) createNamespaces(ctx context.Context, envTypes []environment.Type, user *auth.User, dbTenant *tenant.Tenant) error {
+	clusterNsMapping, err := c.clusterService.GetUserClusterForType(ctx, user)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"err":         err,
+			"tenant":      user.ID,
+			"cluster_url": *user.UserData.Cluster,
```

Comment:
are you sure that `*user.UserData.Cluster` won't panic?

Reply:
Yes, I am. I both cases the user info has to be already fetched (otherwise it would fail earlier)

Reply:
ok, good 👍